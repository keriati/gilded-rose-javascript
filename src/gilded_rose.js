import {
    lensProp,
    over,
    subtract,
    inc,
    propEq,
    dec,
    clamp,
    cond,
    always,
    compose,
    allPass,
    ifElse,
    identity,
} from 'ramda'

export class Item {
    constructor(name, sellIn, quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }
}


export const TICKET_NAME = 'Backstage passes to a TAFKAL80ETC concert'
export const CHEESE_NAME = 'Aged Brie'
export const SULFURAS = 'Sulfuras, Hand of Ragnaros'
export const CONJURED = 'Conjured Mana Cake'

export const QUALITY = 'quality'
export const SELL_IN = 'sellIn'
export const NAME = 'name'

// helpers, utils
export const nameEqual = propEq(NAME)
export const isCheese = nameEqual(CHEESE_NAME)
export const isTicket = nameEqual(TICKET_NAME)
export const isSulfuras = nameEqual(SULFURAS)
export const isConjured = nameEqual(CONJURED)
export const isCommonItem = (item) => !allPass[isConjured, isSulfuras, isCheese, isTicket]

export const qualityLens = lensProp(QUALITY)
export const overQuality = over(qualityLens)

export const qualityBoundaries = clamp(0, 50)
export const increaseQuality = overQuality(inc)
export const decreaseQuality = overQuality(dec)
export const decreaseQualityBy2 = over(qualityLens, subtract(2))
export const boundQuality = overQuality(qualityBoundaries)

export const sellInLens = lensProp(SELL_IN)
export const decreaseSellIn = over(sellInLens, dec)

export const increaseQualityAfterSellDate = (item) => {
    return item.sellIn < 0 ? increaseQuality(item) : item
}
export const decreaseQualityAfterSellDate = (item) => {
    return item.sellIn < 0 ? decreaseQuality(item) : item
}
export const increaseQualityFor10DaysBeforeSellDate = (item) => {
    return item.sellIn < 11 ? increaseQuality(item) : item
}
export const increaseQualityFor6DaysBeforeSellDate = (item) => {
    return item.sellIn < 6 ? increaseQuality(item) : item
}
export const dropQualityToZeroAfterSellDate = (item) => {
    return item.sellIn < 0 ? overQuality(always(0), item) : item
}


// item updates
export const updateCommonItem = compose(
    decreaseQualityAfterSellDate,
    decreaseQuality,
)
export const updateTicket = compose(
    dropQualityToZeroAfterSellDate,
    increaseQualityFor6DaysBeforeSellDate,
    increaseQualityFor10DaysBeforeSellDate,
    increaseQuality
)
export const updateCheese = compose(
    increaseQualityAfterSellDate,
    increaseQuality,
)
export const updateConjuredItem = item => {
   return isConjured(item) ? decreaseQualityBy2(item) : item
}

// general
export const updateQuality = cond([
    [isConjured, updateConjuredItem],
    [isCheese, updateCheese],
    [isTicket, updateTicket],
    [isCommonItem, updateCommonItem],
])

export const update = compose(
    decreaseSellIn,
    boundQuality,
    updateQuality
)

export class Shop {
    constructor(items = []) {
        this.items = items;
    }

    updateQuality() {
        for (let i = 0; i < this.items.length; i++) {
            const item = this.items[i]
            this.items[i] = isSulfuras(item) ? item : update(item)
        }

        return this.items;
    }
}
