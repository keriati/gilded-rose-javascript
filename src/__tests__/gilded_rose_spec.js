import { Shop, Item } from '../gilded_rose';

const MIN_VALUE = 0;
const MAX_VALUE = 50;

describe('Gilded Rose', () => {
    it('creates empty list of items when none is provided', () => {
        const gildedRose = new Shop();

        expect(gildedRose.items.length).toEqual(0);
    });
});

describe('Standard item', () => {
    it('does not change name', () => {
        const gildedRose = new Shop([new Item('foo', 0, 0)]);

        const items = gildedRose.updateQuality();

        expect(items[0].name).toEqual('foo');
    });

    it('decreases sell in date by 1', () => {
        const gildedRose = new Shop([new Item('foo', 1, 1)]);

        const items = gildedRose.updateQuality();

        expect(items[0].sellIn).toEqual(0);
    });

    it('decreases quality by 1 before sell in date', () => {
        const gildedRose = new Shop([new Item('foo', 1, 1)]);

        const items = gildedRose.updateQuality();

        expect(items[0].quality).toEqual(0);
    });

    it('decreases quality by 2 after sell in date', () => {
        const gildedRose = new Shop([new Item('foo', 0, 2)]);

        const items = gildedRose.updateQuality();

        expect(items[0].quality).toEqual(0);
    });

    it('does not decrease quality bellow ' + MIN_VALUE, () => {
        const gildedRose = new Shop([
            new Item('foo', 1, MIN_VALUE),
            new Item('foo', 0, MIN_VALUE + 1)
        ]);

        const items = gildedRose.updateQuality();

        expect(items[0].quality).toEqual(MIN_VALUE);
        expect(items[1].quality).toEqual(MIN_VALUE);
    });
});

describe('Aging item', () => {
    it('increases quality by 1 before sell in date', () => {
        const gildedRose = new Shop([new Item('Aged Brie', 1, 0)]);

        const items = gildedRose.updateQuality();

        expect(items[0].quality).toEqual(1);
    });

    it('increases quality by 2 after sell in date', () => {
        const gildedRose = new Shop([new Item('Aged Brie', 0, 0)]);

        const items = gildedRose.updateQuality();

        expect(items[0].quality).toEqual(2);
    });

    it('does not increase quality above ' + MAX_VALUE, () => {
        const gildedRose = new Shop([
            new Item('Aged Brie', 1, MAX_VALUE),
            new Item('Aged Brie', 0, MAX_VALUE - 1)
        ]);

        const items = gildedRose.updateQuality();

        expect(items[0].quality).toEqual(MAX_VALUE);
        expect(items[1].quality).toEqual(MAX_VALUE);
    });
});

describe('Legendary item', () => {
    it('does not change sell in date', () => {
        const gildedRose = new Shop([new Item('Sulfuras, Hand of Ragnaros', 1, 80)]);

        const items = gildedRose.updateQuality();

        expect(items[0].sellIn).toEqual(1);
    });

    it('does not change quality', () => {
        const gildedRose = new Shop([new Item('Sulfuras, Hand of Ragnaros', 1, 80)]);

        const items = gildedRose.updateQuality();

        expect(items[0].quality).toEqual(80);
    });
});

describe('Backstage passes', () => {
    it('increases quality by 1 more than 10 days before concert', () => {
        const gildedRose = new Shop([new Item('Backstage passes to a TAFKAL80ETC concert', 11, 0)]);

        const items = gildedRose.updateQuality();

        expect(items[0].quality).toEqual(1);
    });
  
    it('increases quality by 2 10 - 6 days before concert', () => {
        const gildedRose = new Shop([
            new Item('Backstage passes to a TAFKAL80ETC concert', 10, 0),
            new Item('Backstage passes to a TAFKAL80ETC concert', 6, 0)
        ]);

        const items = gildedRose.updateQuality();

        expect(items[0].quality).toEqual(2);
        expect(items[1].quality).toEqual(2);
    });

    it('increases quality by 3 5 or less days before concert', () => {
        const gildedRose = new Shop([
            new Item('Backstage passes to a TAFKAL80ETC concert', 5, 0),
            new Item('Backstage passes to a TAFKAL80ETC concert', 1, 0)
        ]);

        const items = gildedRose.updateQuality();

        expect(items[0].quality).toEqual(3);
        expect(items[1].quality).toEqual(3);
    });

    it('decreases quality to 0 after concert', () => {
        const gildedRose = new Shop([new Item('Backstage passes to a TAFKAL80ETC concert', 0, 1)]);

        const items = gildedRose.updateQuality();

        expect(items[0].quality).toEqual(0);
    });
});

describe('Snapshot test', () => {
    it('does not change snapshot', () => {
        const items = [];

        items.push(new Item('+5 Dexterity Vest', 10, 20));
        items.push(new Item('Aged Brie', 2, 0));
        items.push(new Item('Elixir of the Mongoose', 5, 7));
        items.push(new Item('Sulfuras, Hand of Ragnaros', 0, 80));
        items.push(new Item('Sulfuras, Hand of Ragnaros', -1, 80));
        items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 15, 20));
        items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 10, 49));
        items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 5, 49));
        // this conjured item does not work properly yet
        items.push(new Item('Conjured Mana Cake', 3, 6));
        
        
        const gildedRose = new Shop(items);
        
        const days = 2;
        
        for (let i = 0; i < days; i++) {
            gildedRose.updateQuality();
            expect(gildedRose.items).toMatchSnapshot();
        }
        
    });
});