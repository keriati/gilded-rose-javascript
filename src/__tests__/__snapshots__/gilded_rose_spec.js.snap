// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test does not change snapshot 1`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 19,
    "sellIn": 9,
  },
  Item {
    "name": "Aged Brie",
    "quality": 1,
    "sellIn": 1,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 6,
    "sellIn": 4,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 21,
    "sellIn": 14,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 9,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 4,
  },
  Item {
    "name": "Conjured Mana Cake",
    "quality": 5,
    "sellIn": 2,
  },
]
`;

exports[`Snapshot test does not change snapshot 2`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 18,
    "sellIn": 8,
  },
  Item {
    "name": "Aged Brie",
    "quality": 2,
    "sellIn": 0,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 5,
    "sellIn": 3,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 22,
    "sellIn": 13,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 8,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 3,
  },
  Item {
    "name": "Conjured Mana Cake",
    "quality": 4,
    "sellIn": 1,
  },
]
`;
